### PR TITLE
fix: [AA-1087] defer goals widget for proctoring

### PR DESCRIPTION
**Pull request by cdeery**
- pause showing goals widget while proctor info panel awaits callback
AA-1087 Proctor Info Panel

### CODE REVIEW

File: src/course-home/outline-tab/widgets/ProctoringInfoPanel.jsx [hunk 1]
```diff
@@ -82,6 +84,7 @@ function ProctoringInfoPanel({ courseId, username, intl }) {
             if (Object.keys(response).length > 0) {
               setShowInfoPanel(true);
             }
+            isResolved();
```
Comment (Dillon-Dumesnil) [inline src/course-home/outline-tab/widgets/ProctoringInfoPanel.jsx]: My only concern with this PR is will this code path always be hit? What if `getProctoringInfoData` errors? I don't see a `.catch` method, but it seems like in any unsuccessful call, `isResolved` will never resolve. Same with if there isn't a response for some reason. 
Is there a way we can ensure we will always resolve this since it will block the goals widget otherwise.

File: src/course-home/outline-tab/OutlineTab.jsx [hunk 2]
```diff
@@ -49,6 +49,7 @@ function OutlineTab({ intl }) {
 
   const {
     accessExpiration,
+    enableProctoredExams,
```
Comment (Dillon-Dumesnil) [inline src/course-home/outline-tab/OutlineTab.jsx]: Nit: keep these alphabetized. And if you're updating, maybe also update offer to be higher than resumeCourse

File: src/course-home/data/api.js [hunk 3]
```diff
@@ -366,6 +367,7 @@ export async function getOutlineTabData(courseId) {
     datesWidget,
     enrollAlert,
     enrollmentMode,
+    enableProctoredExams,
```
Comment (Dillon-Dumesnil) [inline src/course-home/data/api.js:370]: Nit: keep alphabetized

**Review by Dillon-Dumesnil — COMMENTED** [review]
Couple questions to ensure we resolve the proctoring request.

As a last point too, should we have a test around this? Ensuring goals widget doesn't display until the correct state change?

File: src/course-home/outline-tab/widgets/ProctoringInfoPanel.jsx (hunk 1 above)
Comment (MatthewPiatetsky) [reply src/course-home/outline-tab/widgets/ProctoringInfoPanel.jsx]: agree, i think you might be able to use a .finally( block just for this part as this approach wouldn't show the goals panel if the proctoring request fails

File: src/course-home/outline-tab/OutlineTab.jsx [hunk 4]
```diff
@@ -73,6 +74,7 @@ function OutlineTab({ intl }) {
   const [deprecatedCourseGoalToDisplay, setDeprecatedCourseGoalToDisplay] = useState(selectedGoal);
   const [goalToastHeader, setGoalToastHeader] = useState('');
   const [expandAll, setExpandAll] = useState(false);
+  const [proctorPanelResolved, setProctorPanelResolved] = useState(!enableProctoredExams);
```
Comment (MatthewPiatetsky) [inline src/course-home/outline-tab/OutlineTab.jsx:79]: i think a comment would be worthwhile here to enable what we're doing and why

**Review by MatthewPiatetsky — COMMENTED** [review]
agree that a test could be a good idea since this is very edge casey

**Comment by cdeery** [comment]
I am a little stumped about how to test this. When I mock in a 404 response from the proctoringInfoURL, the test framework fails.
Cannot read property 'httpErrorStatus' of undefined
TypeError: Cannot read property 'httpErrorStatus' of undefined
    at _callee4$ (/Users/cdeery/src/frontend-app-learning/src/course-home/data/api.js:292:13)
    at tryCatch (/Users/cdeery/src/frontend-app-learning/node_modules/axios-cache-adapter/dist/webpack:/axiosCacheAdapter/node_modules/regenerator-runtime/runtime.js:63:15)
    at Generator.invoke [as _invoke] (/Users/cdeery/src/frontend-app-learning/node_modules/axios-cache-adapter/dist/webpack:/axiosCacheAdapter/node_modules/regenerator-runtime/runtime.js:293:1)
    at Generator.throw (/Users/cdeery/src/frontend-app-learning/node_modules/axios-cache-adapter/dist/webpack:/axiosCacheAdapter/node_modules/regenerator-runtime/runtime.js:118:1)
    at asyncGeneratorStep (/Users/cdeery/src/frontend-app-learning/src/course-home/data/api.js:29:103)
    at _throw (/Users/cdeery/src/frontend-app-learning/src/course-home/data/api.js:31:291)

**Comment by MatthewPiatetsky** [comment]
maybe fedx will have suggestions on how to test?
i'm not very familiar with testing in mfes

File: src/course-home/outline-tab/widgets/ProctoringInfoPanel.jsx (hunk 1 above)
Comment (cdeery) [reply src/course-home/outline-tab/widgets/ProctoringInfoPanel.jsx]: Added some tests, it turned out to be tricky because the proctoring info API returns a 404 if the course does not have proctoring info. I modified ProctoringInfoPanel to eat the error. This probably needs a better long term solution, I saw that the API threw 170K 404s into Splunk in 24 hours.

**Comment by cdeery** [comment]
Carla helped me get the testing resolved.

File: src/course-home/outline-tab/OutlineTab.jsx (hunk 4 above)
Comment (Dillon-Dumesnil) [reply src/course-home/outline-tab/OutlineTab.jsx:79]: Nice comment!

File: src/course-home/outline-tab/OutlineTab.test.jsx [hunk 5]
```diff
@@ -529,6 +529,47 @@ describe('Outline Tab', () => {
...
+    const onboardingReleaseDate = new Date();
+    onboardingReleaseDate.setDate(new Date().getDate() - 7);
+
+    beforeEach(async () => {
+      setTabData({
+        course_goals: {
+          weekly_learning_goal_enabled: true,
+        },
+      });
+    });
```
Comment (Dillon-Dumesnil) [inline src/course-home/outline-tab/OutlineTab.test.jsx]: Nit: So there are only 3 tests in this file and you change the value of course_goals in two of them already. I'd say just drop this beforeEach call and add the setTabData into the test itself.

Also, I just realized all of these tests have `weekly_learning_goal_enabled` set to `true`. Do we have any tests where it is false? If not, would you mind adding one? And looking at this now, I realize all of these tests are related to the proctored exam stuff (which fair), but do we have tests for the actual component itself? Do those exist and are just in a different file? Sorry for dumping all of these testing questions on ya last minute.

File: src/course-home/outline-tab/OutlineTab.test.jsx [hunk 6]
```diff
@@ -1172,6 +1213,7 @@ describe('Outline Tab', () => {
 
     it('does not appear for 404', async () => {
       axiosMock.onGet(proctoringInfoUrl).reply(404);
+      await fetchAndRender();
```
Comment (Dillon-Dumesnil) [inline src/course-home/outline-tab/OutlineTab.test.jsx:1211]: 👏

**Review by Dillon-Dumesnil — COMMENTED** [review]
Some questions about tests, but otherwise this PR looks good to me! Great improvements being made here. Nice job!

File: src/course-home/outline-tab/OutlineTab.test.jsx (hunk 5 above)
Comment (cdeery) [reply src/course-home/outline-tab/OutlineTab.test.jsx]: I can drop the beforeEach. It is an artifact of experimentation as I was trying to figure out tests. Good call.
There are tests for weekly learning goal in this file from lines 424 to 531, directly above these tests. You may have misread the diff - there are many more than 3 tests in the file.
The test at line 424 tests that the goal panel is not shown if the the flag is false.

Comment (cdeery) [reply src/course-home/outline-tab/OutlineTab.test.jsx]: You must mean there are only 3 tests in this block. That's fair

Comment (Dillon-Dumesnil) [reply src/course-home/outline-tab/OutlineTab.test.jsx]: Yes, sorry, I meant this block. 
Ah I see the other tests now. Okay, new request because y'boy is picky. So the describe here is "weekly learning goal is displayed". I think if you want to keep that, you should move the other display based tests into this describe. If not, I think the describe should maybe be updated to call out that this section of tests about weekly learning goals is specifically how it interacts with the ProctoringInfoPanel and does not encompass all display behaviors.

Comment (cdeery) [reply src/course-home/outline-tab/OutlineTab.test.jsx]: Folded the proctoring tests into the other goals widget tests

**Review by Dillon-Dumesnil — APPROVED** [review]
LGTM
